File: rider/src/main/java/org/digma/intellij/plugin/rider/psi/csharp/CSharpLanguageService.java
package org.digma.intellij.plugin.rider.psi.csharp;

import com.intellij.codeInsight.codeVision.CodeVisionEntry;
import com.intellij.lang.Language;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.*;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.*;
import com.intellij.openapi.util.TextRange;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.*;
import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent;
import com.jetbrains.rider.projectView.SolutionLifecycleHost;
import kotlin.Pair;
import org.apache.commons.lang3.time.StopWatch;
import org.digma.intellij.plugin.common.*;
import org.digma.intellij.plugin.editor.EditorUtils;
import org.digma.intellij.plugin.errorreporting.ErrorReporter;
import org.digma.intellij.plugin.log.Log;
import org.digma.intellij.plugin.model.discovery.*;
import org.digma.intellij.plugin.psi.*;
import org.digma.intellij.plugin.rider.protocol.*;
import org.digma.intellij.plugin.ui.CaretContextService;
import org.jetbrains.annotations.*;

import java.util.*;
import java.util.concurrent.TimeUnit;

import static org.digma.intellij.plugin.common.CatchingUtilsKt.executeCatchingWithResult;
import static org.digma.intellij.plugin.common.PsiAccessUtilsKt.runInReadAccessWithResult;

public class CSharpLanguageService extends LifetimedProjectComponent implements LanguageService {

    private static final Logger LOGGER = Logger.getInstance(CSharpLanguageService.class);

    private final Project project;


    /*
    It's better, as much as possible, in language services especially, not to initialize service dependencies in the constructor but use
    a getInstance for services when they are first needed. that will minimize the possibility for cyclic dependencies.
     */
    public CSharpLanguageService(Project project) {
        super(project);
        this.project = project;
    }

    @Override
    public void ensureStartupOnEDT(@NotNull Project project) {
        Log.log(LOGGER::debug, "ensureStartupOnEDT called, backend loaded: {}", SolutionLifecycleHost.Companion.getInstance(project).isBackendLoaded().getValue());
        //make sure LanguageServiceHost is initialized on EDT, for example project.solution.languageServiceModel must be
        // called on EDT
        LanguageServiceHost.getInstance(project);
        CodeLensHost.getInstance(project);
    }

    @Override
    public void runWhenSmart(Runnable task) {

        Runnable r = () -> {
            if (DumbService.isDumb(project)) {
                DumbService.getInstance(project).runWhenSmart(task);
            } else {
                task.run();
            }
        };

        LanguageServiceHost.getInstance(project).runIfSolutionLoaded(r);

    }

    @Nullable
    @Override
    public Language getLanguageForMethodCodeObjectId(@NotNull String methodId) {

        if (methodId.indexOf("$_$") <= 0) {
            Log.log(LOGGER::debug, "method id in getLanguageForMethodCodeObjectId does not contain $_$ {}", methodId);
            return null;
        }

        if (LanguageServiceHost.getInstance(project).isCSharpMethod(methodId)) {
            return CSharpLanguageUtil.getCSharpLanguageInstanceWithReflection();
        }
        return null;
    }

    @Nullable
    @Override
    public Language getLanguageForClass(@NotNull String className) {
        //todo: implement
//        if (LanguageServiceHost.getInstance(project).isCSharpClass(className)) {
//            return CSharpLanguageUtil.getCSharpLanguageInstanceWithReflection();
//        }
        return null;
    }

    @Override
    public boolean isSupportedFile(@NotNull Project project, @NotNull VirtualFile newFile) {

        if (!VfsUtilsKt.isValidVirtualFile(newFile)) {
            return false;
        }

        PsiFile psiFile = runInReadAccessWithResult(() -> PsiManager.getInstance(project).findFile(newFile));
        if (!PsiUtils.isValidPsiFile(psiFile)) {
            return false;
        }
        return isSupportedFile(psiFile);
    }

    @Override
    public boolean isSupportedFile(PsiFile psiFile) {
        if (!PsiUtils.isValidPsiFile(psiFile)) {
            return false;
        }
        return CSharpLanguageUtil.isCSharpLanguage(psiFile.getLanguage());
    }


    //CSharpLanguageService needs the Editor to take the projectModelId which is the preferred way to find a IPsiSourceFile in resharper.
    // so try to always send the editor to this method.
    @Override
    @NotNull
    public MethodUnderCaret detectMethodUnderCaret(@NotNull Project project, @NotNull PsiFile psiFile, @Nullable Editor selectedEditor, int caretOffset) {

        //detectMethodUnderCaret should run very fast and return a result,
        // this operation may become invalid very soon if user clicks somewhere else.
        // no retry because it needs to complete very fast
        //it may be called from EDT or background, runInReadAccessWithResult will acquire read access if necessary.
        return executeCatchingWithResult(() -> PsiAccessUtilsKt.runInReadAccessWithResult(() -> LanguageServiceHost.getInstance(project).detectMethodUnderCaret(psiFile, selectedEditor, caretOffset)), throwable -> {
            ErrorReporter.getInstance().reportError(getClass().getSimpleName() + ".detectMethodUnderCaret", throwable);
            return new MethodUnderCaret("", "", "", "", PsiUtils.psiFileToUri(psiFile), caretOffset, null, false);
        });
    }


    @Override
    public void navigateToMethod(String methodId) {

        Log.log(LOGGER::debug, "got navigate to method request {}", methodId);
        if (methodId.indexOf("$_$") <= 0) {
            Log.log(LOGGER::debug, "method id in navigateToMethod does not contain $_$, can not navigate {}", methodId);
            return;
        }

        LanguageServiceHost.getInstance(project).navigateToMethod(methodId);
    }

    @Override
    public boolean isServiceFor(@NotNull Language language) {
        return CSharpLanguageUtil.isCSharpLanguage(language);
    }


    @Override
    public Map<String, String> findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(List<String> codeObjectIds) {
        return ProgressManager.getInstance().
                computeInNonCancelableSection(() -> LanguageServiceHost.getInstance(project).findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(codeObjectIds));
    }

    @Override
    public Map<String, Pair<String, Integer>> findWorkspaceUrisForMethodCodeObjectIds(List<String> methodCodeObjectIds) {
        return ProgressManager.getInstance().
                computeInNonCancelableSection(() -> LanguageServiceHost.getInstance(project).findWorkspaceUrisForMethodCodeObjectIds(methodCodeObjectIds));
    }

    @Override
    public Map<String, Pair<String, Integer>> findWorkspaceUrisForSpanIds(List<String> spanIds) {
        var stopWatch = StopWatch.createStarted();
        try {
            return ProgressManager.getInstance().
                    computeInNonCancelableSection(() -> LanguageServiceHost.getInstance(project).findWorkspaceUrisForSpanIds(spanIds));

        } finally {
            stopWatch.stop();
            Log.log(LOGGER::debug, "findWorkspaceUrisForSpanIds time took {} milliseconds", stopWatch.getTime(TimeUnit.MILLISECONDS));
        }
    }

    @Override
    public Set<EndpointInfo> lookForDiscoveredEndpoints(String endpointId) {
        return Collections.emptySet();
    }

    @Override
    public void environmentChanged(String newEnv, boolean refreshInsightsView) {

        if (refreshInsightsView) {
            EDT.ensureEDT(() -> {
                var fileEditor = FileEditorManager.getInstance(project).getSelectedEditor();
                if (fileEditor != null) {
                    var file = fileEditor.getFile();
                    if (VfsUtilsKt.isValidVirtualFile(file)) {
                        var psiFile = PsiManager.getInstance(project).findFile(file);
                        if (PsiUtils.isValidPsiFile(psiFile) && isRelevant(psiFile.getVirtualFile())) {
                            var selectedTextEditor = EditorUtils.getSelectedTextEditorForFile(file, FileEditorManager.getInstance(project));
                            if (selectedTextEditor != null) {
                                int offset = selectedTextEditor.getCaretModel().getOffset();
                                var methodUnderCaret = detectMethodUnderCaret(project, psiFile, selectedTextEditor, offset);
                                CaretContextService.getInstance(project).contextChanged(methodUnderCaret);
                            }
                        }
                    }
                }
            });
        }


        CodeLensHost.getInstance(project).environmentChanged(newEnv);
    }


    @Override
    public @NotNull DocumentInfo buildDocumentInfo(@NotNull PsiFile psiFile, BuildDocumentInfoProcessContext context) {
        return buildDocumentInfo(psiFile, null, context);
    }

    @Override
    public @NotNull DocumentInfo buildDocumentInfo(@NotNull PsiFile psiFile, @Nullable FileEditor newEditor, BuildDocumentInfoProcessContext context) {

        Log.log(LOGGER::debug, "got buildDocumentInfo request for {}", psiFile);
        //must be PsiJavaFile , this method should be called only for java files
        if (CSharpLanguageUtil.isCSharpFile(psiFile)) {
            DocumentInfo documentInfo = LanguageServiceHost.getInstance(project).getDocumentInfo(psiFile, newEditor);
            if (documentInfo == null) {
                Log.log(LOGGER::warn, "DocumentInfo not found for {}, returning empty DocumentInfo", psiFile);
                documentInfo = new DocumentInfo(PsiUtils.psiFileToUri(psiFile), new HashMap<>());
            }
            return documentInfo;
        } else {
            Log.log(LOGGER::debug, "psi file is noy CSharpFile, returning empty DocumentInfo for {}", psiFile);
            return new DocumentInfo(PsiUtils.psiFileToUri(psiFile), new HashMap<>());
        }
    }


    @Override
    public boolean isRelevant(VirtualFile file) {
        if (file.isDirectory()) {
            return false;
        }

        PsiFile psiFile = PsiManager.getInstance(project).findFile(file);
        if (!PsiUtils.isValidPsiFile(psiFile)) {
            return false;
        }

        return isRelevant(psiFile);

    }

    @Override
    public boolean isRelevant(PsiFile psiFile) {
        return ReadActions.ensureReadAction(() ->
                PsiUtils.isValidPsiFile(psiFile) &&
                psiFile.isWritable() &&
                        isSupportedFile(psiFile));

    }

    @Override
    public void refreshMethodUnderCaret(@NotNull Project project, @NotNull PsiFile psiFile, @Nullable Editor selectedEditor, int offset) {
        MethodUnderCaret methodUnderCaret = detectMethodUnderCaret(project, psiFile, selectedEditor, offset);
        CaretContextService.getInstance(project).contextChanged(methodUnderCaret);
    }

    @Override
    public boolean isCodeVisionSupported() {
        return false;
    }

    @Override
    public @NotNull List<Pair<TextRange, CodeVisionEntry>> getCodeLens(@NotNull PsiFile psiFile) {
        throw new UnsupportedOperationException("should not be called for CSharpLanguageService");
    }

    @Override
    public @Nullable PsiElement getPsiElementForMethod(@NotNull String methodId) {
        //todo: implement
        return null;
    }

    @Override
    public @Nullable PsiElement getPsiElementForClassByMethodId(@NotNull String methodId) {
        return null;
    }

    @Override
    public @Nullable PsiElement getPsiElementForClassByName(@NotNull String className) {
        return null;
    }
}

File: rider/src/main/kotlin/org/digma/intellij/plugin/rider/psi/csharp/CSharpLanguageService.kt
package org.digma.intellij.plugin.rider.psi.csharp

import com.intellij.lang.Language
import com.intellij.openapi.application.EDT
import com.intellij.openapi.application.smartReadAction
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import com.jetbrains.rd.util.reactive.whenTrue
import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
import com.jetbrains.rider.languages.fileTypes.csharp.CSharpFileType
import com.jetbrains.rider.languages.fileTypes.csharp.CSharpLanguage
import com.jetbrains.rider.projectView.SolutionLifecycleHost
import com.jetbrains.rider.projectView.SolutionStartupService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.digma.intellij.plugin.common.EDT
import org.digma.intellij.plugin.common.ReadActions
import org.digma.intellij.plugin.common.isProjectValid
import org.digma.intellij.plugin.common.isValidVirtualFile
import org.digma.intellij.plugin.common.suspendableRetry
import org.digma.intellij.plugin.discovery.model.EndpointLocation
import org.digma.intellij.plugin.document.DocumentInfoStorage
import org.digma.intellij.plugin.document.getSelectedTextEditorForFile
import org.digma.intellij.plugin.errorreporting.ErrorReporter
import org.digma.intellij.plugin.instrumentation.InstrumentationProvider
import org.digma.intellij.plugin.instrumentation.NoOpInstrumentationProvider
import org.digma.intellij.plugin.log.Log
import org.digma.intellij.plugin.model.discovery.DocumentInfo
import org.digma.intellij.plugin.model.discovery.MethodUnderCaret
import org.digma.intellij.plugin.psi.LanguageService
import org.digma.intellij.plugin.psi.PsiUtils
import org.digma.intellij.plugin.rider.protocol.LanguageServiceHost
import kotlin.time.Duration.Companion.seconds

@Suppress("LightServiceMigrationCode")
class CSharpLanguageService(project: Project) : LifetimedProjectComponent(project), LanguageService {

    private val logger = Logger.getInstance(this::class.java)


    init {
        SolutionLifecycleHost.getInstance(project).isBackendLoaded.whenTrue(componentLifetime) {

            /*
            Idea and Rider may reopen files on startup.
            In Idea, we wait for smart mode on startup before building DocumentInfo for open files.
            In Rider the indexes are not ready when building the DocumentInfo for open files.
            Here we wait for the solution loaded and simulate fileOpened again for every file in DocumentInfoStorage
            This code runs on EDT and should not block EDT for a long time.
             */

            Log.trace(
                logger, "Rider solution loaded, warm startup: {}, initializing {}",
                SolutionStartupService.getInstance(project).isWarmStartup(), SolutionStartupService.getInstance(project).isInitializing()
            )

            FileEditorManager.getInstance(project).openFiles.forEach { file ->
                componentLifetime.coroutineScope.launch {
                    delay(1.seconds) //just wait a second for the solution to fully load, it seems necessary
                    Log.trace(logger, "Rider solution loaded, building DocumentInfo for {}", file)
                    val documentInfo = buildDocumentInfo(file)
                    if (documentInfo != null) {
                        Log.trace(logger, "Rider solution loaded, done building DocumentInfo for {}", file)
                        DocumentInfoStorage.getInstance(project).updateDocumentInfoForRider(file, documentInfo)
                    } else {
                        Log.log(logger::warn, "Rider solution loaded,Could not build DocumentInfo for {}", file)
                    }
                }
            }
        }
    }


    override fun getLanguage(): Language {
        return CSharpLanguage
    }

    override fun getFileType(): FileType {
        return CSharpFileType
    }

    override suspend fun getLanguageForMethodCodeObjectId(methodId: String): Language? {

        if (methodId.indexOf("\$_$") <= 0) {
            Log.trace(logger, "method id in getLanguageForMethodCodeObjectId does not contain \$_$ {}", methodId)
            return null
        }

        return if (LanguageServiceHost.getInstance(project).isCSharpMethod(methodId)) {
            return CSharpLanguage
        } else {
            null
        }
    }


    override suspend fun getLanguageForClass(className: String): Language? {
        //todo: implement
//        if (LanguageServiceHost.getInstance(project).isCSharpClass(className)) {
//            return CSharpLanguageUtil.getCSharpLanguageInstanceWithReflection();
//        }
        return null
    }


    override fun isSupportedFile(psiFile: PsiFile): Boolean {
        return CSharpLanguage == psiFile.language
    }


    override suspend fun detectMethodUnderCaret(virtualFile: VirtualFile, editor: Editor, caretOffset: Int): MethodUnderCaret {
        return LanguageServiceHost.getInstance(project).detectMethodUnderCaret(virtualFile, editor, caretOffset)
    }


    override suspend fun detectMethodBySpan(project: Project, spanCodeObjectId: String): String? {
        return try {
            LanguageServiceHost.getInstance(project).getMethodIdBySpanId(spanCodeObjectId)
        } catch (e: Throwable) {
            Log.warnWithException(logger, project, e, "Error in detectMethodBySpan {}", e)
            ErrorReporter.getInstance().reportError(project, javaClass.getSimpleName() + ".detectMethodBySpan", e)
            null
        }
    }


    override suspend fun navigateToMethod(methodId: String) {

        Log.trace(logger, "got navigate to method request {}", methodId)
        if (methodId.indexOf("\$_$") <= 0) {
            Log.trace(logger, "method id in navigateToMethod does not contain \$_$, can not navigate {}", methodId)
            return
        }

        LanguageServiceHost.getInstance(project).navigateToMethod(methodId)
    }


    override suspend fun findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(methodCodeObjectIds: List<String>): Map<String, String> {
        return LanguageServiceHost.getInstance(project).findWorkspaceUrisForCodeObjectIdsForErrorStackTrace(methodCodeObjectIds)
    }

    override suspend fun findWorkspaceUrisForMethodCodeObjectIds(methodCodeObjectIds: List<String>): Map<String, Pair<String, Int>> {
        return LanguageServiceHost.getInstance(project).findWorkspaceUrisForMethodCodeObjectIds(methodCodeObjectIds)
    }


    override suspend fun findWorkspaceUrisForSpanIds(spanIds: List<String>): Map<String, Pair<String, Int>> {
        return LanguageServiceHost.getInstance(project).findWorkspaceUrisForSpanIds(spanIds)
    }

    override fun lookForDiscoveredEndpoints(endpointId: String): Set<EndpointLocation> {
        return emptySet()
    }

    /**
     * Builds a DocumentInfo.
     * This method doesn't handle exceptions and may throw any kind of exception
     */
    override suspend fun buildDocumentInfo(virtualFile: VirtualFile): DocumentInfo? {

        //don't call this method on EDT or in read access, read access is taken when needed
        EDT.assertNonDispatchThread()
        ReadActions.assertNotInReadAccess()

        Log.trace(logger, project, "got buildDocumentInfo request for {}", virtualFile)

        if (!isValidVirtualFile(virtualFile)) {
            Log.trace(logger, project, "buildDocumentInfo: virtualFile is not valid for {}", virtualFile)
            return null
        }

        //we don't need the psi file here, discovery is done in Rider backend. this is just a validation check
        // to make sure that it's a csharp file that can still be found
        val psiFile = smartReadAction(project) {
            val psiFile = PsiManager.getInstance(project).findFile(virtualFile)
            if (psiFile == null) {
                Log.trace(logger, project, "buildDocumentInfo: could not find psiFile for {}", virtualFile)
                null
            } else if (!PsiUtils.isValidPsiFile(psiFile)) {
                Log.trace(logger, project, "buildDocumentInfo: psiFile is not valid for {}", virtualFile)
                null
            } else if (!isSupportedFile(psiFile)) {
                Log.trace(logger, project, "buildDocumentInfo: psiFile is not supported for {}", virtualFile)
                null
            } else if (!isProjectValid(project)) {
                Log.trace(logger, project, "buildDocumentInfo: project is not valid for {}", virtualFile)
                null
            } else {
                psiFile
            }
        }

        return psiFile?.let {
            suspendableRetry {

                //In Rider the preferred way to find a psi source file in the backend is to use projectModelId which can be found
                //  in the frontend editor. If there is no editor, the psi will be searched by the file url
                //todo: this is a costly operation because it must run on EDT. it happens on every file open and every time
                // the DocumentInfo needs to update. consider removing it or refactor Rider backend to do something else.
                val editor: Editor? = withContext(Dispatchers.EDT) {
                    getSelectedTextEditorForFile(project, virtualFile)
                }

                LanguageServiceHost.getInstance(project).getDocumentInfo(virtualFile, editor, getLanguage())
            }
        }
    }


    override fun isCodeVisionSupported(): Boolean {
        return false
    }


    override suspend fun findMethodPsiElementByMethodId(methodId: String): PsiElement? {
        //not relevant for Rider
        return null
    }

    override suspend fun findClassPsiElementByMethodId(methodId: String): PsiElement? {
        //this method is used for running tests, not supported in Rider
        return null
    }

    override suspend fun findClassPsiElementByClassName(className: String): PsiElement? {
        //this method is used for running tests, not supported in Rider
        return null
    }

    override fun getInstrumentationProvider(): InstrumentationProvider {
        return NoOpInstrumentationProvider()
    }

    override fun findMethodsByCodeObjectIds(psiFile: PsiFile, methodIds: List<String>): Map<String, PsiElement> {
        //Never called for csharp language,
        //this method is used by CodeLensService, which is only relevant for java/kotlin/python.
        //Rider does code lens differently in Rider backend, so this method is never called in Rider.
        throw UnsupportedOperationException("should not be called for CSharp")
    }

    override fun dispose() {

    }
}

